flake.nix

```diff
@@ -27,7 +27,16 @@
       systems = lib.intersectLists lib.systems.flakeExposed lib.platforms.linux;
 
       forAllSystems = lib.genAttrs systems;
-      nixpkgsFor = forAllSystems (system: nixpkgs.legacyPackages.${system});
+      nixpkgsFor = forAllSystems (
+        system:
+        import nixpkgs {
+          inherit system;
+          overlays = [
+            self.overlays.default
+            rust-overlay.overlays.rust-overlay
+          ];
+        }
+      );
```

Comment:
Don't instantiate our own Nixpkgs here. There isn't any real benefit from doing so besides making things more "uniform" (FYI the current method of using the overlay as a function is [quite well known](https://jade.fyi/blog/flakes-arent-real/) and not "incorrect"), but there are [massive costs for consumers of this Flake](https://zimbatm.com/notes/1000-instances-of-nixpkgs). Specifically, this adds 100mb+ and a few seconds to any given evaluation

***This is a footgun and was avoided purposefully***

Reply:
ohh...

yeah, so i've read [that article](https://zimbatm.com/notes/1000-instances-of-nixpkgs) before but apparently not closely enough. my takeaway was about storage space. wow.

this explains why my system takes ages to eval. i think i've got some spring cleaning to do. this is surely not the only reason, but i do make extensive use of overlays which i can't imagine helps.

Reply:
> FYI the current method of using the overlay as a function is [,.,] not "incorrect"

i mean yeah. overlay is just a function and since we weren't using `prev:` it's whatever, but i still thought it looked very ugly and bad to be passing `null` where it expects an instance of nixpkgs. i've refactored the code to move the package definition higher up, and not rely on the overlay at all internally, so we can just call `pkgs.callPackage` instead. this avoids the need to even consider how best to use the overlay here; because the overlay isn't for internal use in this flake.